dds/DCPS/XTypes/TypeObject.h

```diff
@@ -826,6 +826,7 @@ namespace XTypes {
 
   // ID of a type member
   typedef ACE_CDR::ULong MemberId;
+  const ACE_CDR::ULong MEMBER_ID_INVALID = UINT32_MAX;
```

Comment:
Does this need to be checked in the IDL compiler to make sure it isn't actually used as an ID?

Reply:
> 7.2.2.4.4.4.4 Member IDs
> Each member of an aggregated type is uniquely identified within that type by an integer
> “member ID.” Member IDs are unsigned and have a range that can be represented in 28 bits:
> from zero to 268,435,455 (0x0FFFFFFF). (The full range of a 32-bit unsigned integer is not used
> in order to allow binary Data Representations the freedom to embed a small amount of meta-data
> into a single 32-bit field if they so desire.)

> The remaining part of the member ID range—from 0 to 268,402,687 (0x0FFFBFFF)—is
> available for use by application-defined types compliant with this specification.

According to the spec, UINT32_MAX will never be naturally used as a MemberID.

The only possible collision would occur with "binary Data Representations", but I have no clue what the spec is talking about with that.

Reply:
> The only possible collision would occur with "binary Data Representations", but I have no clue what the spec is talking about with that.

It's talking about XCDR2 where there is the must understand flag and the length code in the 4 most significant bits of the parameter id, with the rest of the 28 bits being for the member id.

> Does this need to be checked in the IDL compiler to make sure it isn't actually used as an ID?

It doesn't look like we do:

https://github.com/objectcomputing/OpenDDS/blob/7a16ebb3052cb05473a5899612576a35799876c9/dds/idl/be_global.cpp#L988-L1014

`hash_member_name_to_id` won't break 28 bits, so I think the only thing to worry about at the moment would be if a user did something like `@id(0xffffffff)`.

To be safe we should probably add checks for this to `compute_id` and `write_parameter_id`.